mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdToImapUidDAO.java

```diff
@@ -164,23 +166,18 @@ private PreparedStatement prepareSelect(Session session) {
                 .setSet(USER_FLAGS, ImmutableSet.copyOf(flags.getUserFlags())));
     }
 
-    public CompletableFuture<Boolean> updateMetadata(ComposedMessageIdWithMetaData composedMessageIdWithMetaData, long oldModSeq) {
-        ComposedMessageId composedMessageId = composedMessageIdWithMetaData.getComposedMessageId();
-        Flags flags = composedMessageIdWithMetaData.getFlags();
-        return cassandraAsyncExecutor.executeReturnApplied(update.bind()
-                .setLong(MOD_SEQ, composedMessageIdWithMetaData.getModSeq())
-                .setBool(ANSWERED, flags.contains(Flag.ANSWERED))
-                .setBool(DELETED, flags.contains(Flag.DELETED))
-                .setBool(DRAFT, flags.contains(Flag.DRAFT))
-                .setBool(FLAGGED, flags.contains(Flag.FLAGGED))
-                .setBool(RECENT, flags.contains(Flag.RECENT))
-                .setBool(SEEN, flags.contains(Flag.SEEN))
-                .setBool(USER, flags.contains(Flag.USER))
-                .setSet(USER_FLAGS, ImmutableSet.copyOf(flags.getUserFlags()))
-                .setUUID(MESSAGE_ID, ((CassandraMessageId) composedMessageId.getMessageId()).get())
-                .setUUID(MAILBOX_ID, ((CassandraId) composedMessageId.getMailboxId()).asUuid())
-                .setLong(IMAP_UID, composedMessageId.getUid().asLong())
-                .setLong(MOD_SEQ_CONDITION, oldModSeq));
+    public CompletableFuture<Boolean> updateMetadata(ComposedMessageIdWithMetaData composedMessageIdWithMetaData, FlagsUpdateCalculator flagsUpdateCalculator) {
+        CassandraId mailboxId = (CassandraId) composedMessageIdWithMetaData.getComposedMessageId().getMailboxId();
+        CassandraMessageId messageId = (CassandraMessageId) composedMessageIdWithMetaData.getComposedMessageId().getMessageId();
+
+        return cassandraAsyncExecutor.executeVoid(
+            FlagsUpdateHelper.updateFlags(flagsUpdateCalculator)
+                .apply(update(TABLE_NAME)
+                    .with(set(MOD_SEQ, composedMessageIdWithMetaData.getModSeq())))
+                .where(eq(MAILBOX_ID, mailboxId.asUuid()))
+                .and(eq(MESSAGE_ID, messageId.get()))
+                .and(eq(IMAP_UID, composedMessageIdWithMetaData.getComposedMessageId().getUid().asLong())))
+            .thenApply(any -> true);
```

Comment:
Well, even if the code between the two DAOs are close, I don't think it's desirable to factorise more...

I already extracted common shared fields (flags). I don't want to factorize code on non shared properties.